### PR TITLE
DOC: explain too many ticks

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Users _often_ get confused when they get too many ticks or those that are out of order, because we plot lists of strings as categoricals.  This PR seeks to at least have a FAQ entry to address that problem and encourage them to convert to numbers of some sort.

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by story645** [comment]
I think it's a good idea in theory but dunno that the FAQ (which is hard to find & I think @timhoffm has been moving stuff out of) is the best place for it. Is there a troubleshooting section on one of the date tutorials?

**Comment by jklymak** [comment]
It could go in troubleshooting versus how_to, for sure.  I do not think "examples" or "tutorials" is a good place for this in particular - though we certainly should point out trouble shooting issues in the relevant place _as well_.  

Certainly if I type "matplotlib faq" into google I get https://matplotlib.org/stable/users/faq/index.html as the first entry...

I also don't agree that our users guide should be completely empty, and that we should move _everything_ to tutorials.  Probably somewhat the opposite - we could consider the tutorials the rough draft of what should go in the proper user manual.

**Comment by timhoffm** [comment]
We don't have a proper category for this currently. It's not a "How-to", it's also not "Troubleshooting", which is only technical stuff like finding out your version and reporting a bug.

I can see the need for real FAQ-style "Matplotlib behaves unexpectedly" (better name t.b.d.). That could contain a couple of common errors that are otherwise hard to document: Aside from this:
- "Why are positions of artists not correct when I read them" --> need to draw/layout fist
- "Why do you have multiple ways to do the same thing (APIs)
- "Why does my figure not show" (if we can give helpful information to this general problem.

**Comment by jklymak** [comment]
We could have a "Common Issues" page/section.  OTOH, I think this fits just as well in the "how to" section eg "How to fix this problem..."  I'd personally prefer a one-stop page for all the things I never remember (lets add setting tick label properties to the list, which I have to look up _every_ time, and the right way is always last in whatever I find in stackoverflow)

However, I could easily imagine this "how-to" page getting out of hand and needing its own sub headings.

**Comment by jklymak** [comment]
Newest commit adds a section on artist extents.  

BTW, I'm not sure what to do about "Why does my figure not show".  Usually boils down to "you are on a headless machine" doesn't it?

**Comment by jklymak** [comment]
New TOC: https://65496-1385122-gh.circle-artifacts.com/0/doc/build/html/users/faq/index.html

**Comment by jklymak** [comment]
Can I ping for this PR and ask that any renaming/reorg get moved to an issue?

File: doc/users/faq/howto_faq.rst [hunk 1]
```diff
@@ -9,6 +9,80 @@ How-to
...
+(:doc:`/gallery/lines_bars_and_markers/categorical_variables`), and by default
+puts one tick per "category", and plots them in the order in which they are
+supplied.
+
+In the example below, the upper row plots are plotted using strings for *x*;
+note that each string gets a tick, and they are in the order of the list passed
+to Matplotlib. In the lower row the data is converted to either floats or
+datetime64; note that the ticks are now ordered and spaced numerically.
+
+.. plot::
```
Comment (timhoffm) [inline doc/users/faq/howto_faq.rst:32]: I find the examples a bit extensive. Would the following be sufficient?

```
import matplotlib.pyplot as plt
import numpy as np

fig, ax = plt.subplots(1, 2, constrained_layout=True, figsize=(6, 2))

ax[0].set_title('Ticks seem out of order / misplaced')
x = ['5', '20', '1', '9']  # strings
y = [5, 20, 1, 9]
ax[0].plot(x, y, 'd')
ax[0].tick_params(axis='x', labelcolor='red', labelsize=14)

ax[1].set_title('Many ticks')
x = [str(xx) for xx in np.arange(100)]  # strings
y = np.arange(100)
ax[1].plot(x, y)
ax[1].tick_params(axis='x', labelcolor='red', labelsize=14)
```

grafik

File: doc/users/faq/howto_faq.rst [hunk 2]
```diff
@@ -9,6 +9,80 @@ How-to
...
+
+Determine the extent of Artists on the Figure
+---------------------------------------------
+
+Sometimes we want to know the extent of an Artist.  Matplotlib `.Artist` objects
+have a method `.Artist.get_window_extent` that will usually return the extent of
+the artist in pixels.  However, some artists, in particular text, must be
+rendered at least once before their extent is known.  Matplotlib supplies
+`.Figure.draw_without_rendering`, which should be called before calling
+``get_window_extent``.
```
[the remark this answers is not in the document]
Comment (jklymak) [reply doc/users/faq/howto_faq.rst:83]: I guess I largely know of texts.  I guess there may be some artists that are proportional to the figure or axes size?

File: doc/users/faq/howto_faq.rst (hunk 1 above)
Comment (timhoffm) [reply doc/users/faq/howto_faq.rst:31]: I see that you have updated the examples, and they are more clear now.

Still, do we need more than a screen full of figures and code to illustrate the relatively simple fact "passing string data plots all values as ticks in the given order"?

Motivation for the proposed plot above:
- I've limited this to the error cases. Plotting them is helpful to see how strings typically show up as "wrong" itics
- I've left out the respective "correct" plots with numbers for brevity. It should be quite clear from highlighted ticks that these are undesired. How the correct plot looks does not add much information IMHO.
- I've chosen two plots which show the characteristic error signatures "wrong order" and "too many ticks"
  "Dates out of order" is essentially the same as "numbers out of order".

I favor this to reduce clutter and keep the information density high. But I won't fight over it. If you feel strongly on having the correct plot as well, I suggest to at least remove the dates and use the same layout in both plots to make them equal in size.

File: doc/users/faq/howto_faq.rst [hunk 3]
```diff
@@ -9,6 +9,95 @@ How-to
...
+when reading in a comma-delimited text file. Matplotlib treats lists of strings
+as "categorical" variables
+(:doc:`/gallery/lines_bars_and_markers/categorical_variables`), and by default
+puts one tick per "category", and plots them in the order in which they are
+supplied.
+
+In the example below, the upper row plots are plotted using strings for *x*;
+note that each string gets a tick, and they are in the order of the list passed
+to Matplotlib.  If this is not desired, we need to change *x* to an array of
+numbers.
```
Comment (timhoffm) [inline doc/users/faq/howto_faq.rst:29]: Needs updating for the changed example code below.

File: doc/users/faq/howto_faq.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (timhoffm) [reply doc/users/faq/howto_faq.rst:99]: I feel that vague statements are rather confusing to the user. We should do either

- dig into this deeper and explicitly write out for which artists this is necessary. - This is some work.
- State that users should always call `draw_without_rendering` before using `get_window_extent()`. - I'm a bit uneasy with that because we're forcing additional code and overhead on the user also in cases that would work without it.

I suggest to move this to a separate PR to not block the ticks stuff.

Note also that #22122 now mentions `draw_without_rendering()` in the `Text.get_window_extent()` exception, so that discussing `draw_without_rendering()` here for that case is less important now. If the error happens, the user has the solution at hand.

File: doc/users/faq/howto_faq.rst (hunk 1 above)
Comment (jklymak) [reply doc/users/faq/howto_faq.rst:31]: My apologies, I wrote a reply, but must have closed it before hitting save....

a) datetime-as-string is a _very_ common failure mode, so I think it should be in the FAQ
b) for beginners "turn your strings into floats" is going to require a google search;  half the time, even I have to look up how to do datetimes.  I felt it was important to offer a basic solution, and also to show the "expected" output so the "unexpected" is crystal clear. 

This section is for folks who are trying to do their first plot, reading their Excel data in as a csv.  While I agree we should be as concise as possible, I think here the extra guidance is worth the virtual real estate to get them on the right path.

File: doc/users/faq/howto_faq.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (jklymak) [reply doc/users/faq/howto_faq.rst:99]: I don't think people are coming to a "How to" for the exact reason their artist extent is wrong - they just want to fix it.  I don't think we need to enumerate all the troublesome artists, or interactions with layout managers etc here - we just want to tell them to call a  fake draw and see if that helps.

[the remark this answers is not in the document]
Comment (timhoffm) [reply doc/users/faq/howto_faq.rst:99]: Ok.

File: doc/users/faq/howto_faq.rst (hunk 1 above)
Comment (timhoffm) [reply doc/users/faq/howto_faq.rst:31]: > This section is for folks who are trying to do their first plot, reading their Excel data in as a csv.

IMHO this can happen with all levels of experience.

Anyway, this is a judgement call on the audience and the amount of detail we think is necessary. I appreciate your position, but cannot sign-off on it. If others approve and merge I'm fine with that.

Comment (jklymak) [reply doc/users/faq/howto_faq.rst:31]: I mean its a How-To - I would argue we should provide concrete solutions.

Comment (timhoffm) [reply doc/users/faq/howto_faq.rst:31]: It‘s not really a how-to, it‘s only in how-to because we don‘t have the proper category yet. https://github.com/matplotlib/matplotlib/pull/21943#issuecomment-994113006

Comment (jklymak) [reply doc/users/faq/howto_faq.rst:31]: Let me rephrase: whatever we call it, I think it should have a solution in it, not just the problem.

Comment (timhoffm) [reply doc/users/faq/howto_faq.rst:31]: To me the solution is „don‘t pass in strings“. How to get from strings to numerical values depends on the data, and is not Matplotlibs business. I‘m fine that some people may have to do an additional google search. As said, it‘s a judgement call.

Comment (jklymak) [reply doc/users/faq/howto_faq.rst:26]: OK, well, lets move the full example into an example.

File: doc/users/faq/howto_faq.rst [hunk 4]
```diff
@@ -9,6 +9,61 @@ How-to
...
+    x = [str(xx) for xx in np.arange(100)]  # strings
+    y = np.arange(100)
+    ax[1].plot(x, y)
+    ax[1].tick_params(axis='x', labelcolor='red', labelsize=14)
+
+The solution is to convert the list of strings to numbers or
+datetime objects (often ``np.asarray(['2', '5', '1'], dtype='float')`` or::
+
+    np.asarray(['2021-10-01', '2021-11-02', '2021-12-03'],
+               dtype='datetime64[s]')
```
Comment (timhoffm) [inline doc/users/faq/howto_faq.rst:51]: Let's keep the formatting of both symmetric (not one inline and one as code block). Also it may make it more clear when using a descriptive variable name instead of example data.

```suggestion
datetime objects (often ``np.asarray(numeric_strings, dtype='float')`` or
``np.asarray(datetime_strings, dtype='datetime64[s]')``).
```

File: doc/users/faq/howto_faq.rst [hunk 5]
```diff
@@ -9,6 +9,61 @@ How-to
...
    :backlinks: none
 
+
+.. _how-to-too-many-ticks:
+
+Why do I have so many ticks, and/or why are they out of order?
+--------------------------------------------------------------
+
+One common cause for unexpected tick behavior is passing a list of strings
+instead of numbers or datetime objects. This can easily happen without notice
```
Comment (timhoffm) [inline doc/users/faq/howto_faq.rst:19]: Optional: Should we markup the reason? This could help when quickly skimming the explantion.

```suggestion
One common cause for unexpected tick behavior is passing a *list of strings
instead of numbers or datetime objects*. This can easily happen without notice
```
Alternatively even bold?

File: doc/users/faq/howto_faq.rst [hunk 6]
```diff
@@ -9,6 +9,61 @@ How-to
...
+when reading in a comma-delimited text file. Matplotlib treats lists of strings
+as "categorical" variables
```
Comment (timhoffm) [inline doc/users/faq/howto_faq.rst:21]: I think italics is is more suited than quotes. Italics are often used for technical terms or emphasis, both of which intentions work here. Quotes are used for [many things](https://thevisualcommunicationguy.com/2013/09/11/10-things-you-really-need-to-know-about-quotation-marks/) but IMHO nothing applies here.

```suggestion
as *categorical* variables
```

File: doc/users/faq/howto_faq.rst [hunk 7]
```diff
@@ -9,6 +9,61 @@ How-to
...
+(:doc:`/gallery/lines_bars_and_markers/categorical_variables`), and by default
+puts one tick per "category", and plots them in the order in which they are
```
Comment (timhoffm) [inline doc/users/faq/howto_faq.rst:23]:
```suggestion
puts one tick per category, and plots them in the order in which they are
```
Or put italics if you want.

File: examples/ticks/ticks_too_many.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
+"""
+=====================
+Fixing too many ticks
+=====================
+
+One common cause for unexpected tick behavior is passing a list of strings
+instead of numbers or datetime objects. This can easily happen without notice
+when reading in a comma-delimited text file. Matplotlib treats lists of strings
+as "categorical" variables
```
Comment (timhoffm) [inline examples/ticks/ticks_too_many.py:9]:
```suggestion
as *categorical* variables
```
as above

File: examples/ticks/ticks_too_many.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+(:doc:`/gallery/lines_bars_and_markers/categorical_variables`), and by default
+puts one tick per "category", and plots them in the order in which they are
```
Comment (timhoffm) [inline examples/ticks/ticks_too_many.py:11]:
```suggestion
puts one tick per category, and plots them in the order in which they are
```

as above

File: examples/ticks/ticks_too_many.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+supplied.  If this is not desired, the solution is to convert the strings to
+a numeric type as in the following examples.
+
+"""
+
```
Comment (timhoffm) [inline examples/ticks/ticks_too_many.py:16]: Using sections for the examples structures the content and announces what you'll see in that code blocki

```suggestion

############################################################################
# Example 1: Strings can lead to an unexpected order of number ticks
# ------------------------------------------------------------------

```

File: examples/ticks/ticks_too_many.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+ax[0].set_xlabel('Categories')
+ax[0].set_title('Ticks seem out of order / misplaced')
+
+# convert to numbers:
+x = np.asarray(x, dtype='float')
+ax[1].plot(x, y, 'd')
+ax[1].set_xlabel('Floats')
+ax[1].set_title('Ticks as expected')
+
+############################################################################
```
Comment (timhoffm) [inline examples/ticks/ticks_too_many.py:34]:
```suggestion
############################################################################
# Example 2: Strings can lead to very many ticks
# ----------------------------------------------
```

File: examples/ticks/ticks_too_many.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+ax[0].plot(x, y)
+ax[0].tick_params(axis='x', color='r', labelcolor='r')
+ax[0].set_title('Too many ticks')
+ax[0].set_xlabel('Categories')
+
+ax[1].plot(np.asarray(x, float), y)
+ax[1].set_title('x converted to numbers')
+ax[1].set_xlabel('Floats')
+
+############################################################################
```
Comment (timhoffm) [inline examples/ticks/ticks_too_many.py:50]:
```suggestion
############################################################################
# Example 3: Strings can lead to an unexpected order of datetime ticks
# --------------------------------------------------------------------
```

**Review by timhoffm — COMMENTED** [review]
I :+1: on this conceptually now. Only some minor cosmetic remarks.

**Review by timhoffm — APPROVED** [review]

**Comment by jklymak** [comment]
Thanks for your help @timhoffm !

**Comment by timhoffm** [comment]
Thanks for enduring my review @jklymak! :smile:
